feat(curriculum): explain shorthand stat calculator (#55846)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-advanced-array-methods-by-building-a-statistics-calculator/66bbb13f90639e7a6d664153.md

---
id: 66bbb13f90639e7a6d664153
title: Step 34
challengeType: 0
dashedName: step-34
---

# --description--

While `getMode` is looking great, the `forEach` loop can simplified further. In order to do so, the function will be turned into an `arrow function expression`. This kind of expression is known for not using curly braces and shorter syntax. 

An example of such an expression is the following: 

```js
const numbersArr = [1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4];
const counts = {};

> numbersArr.forEach(el => counts[el] = (counts[el] || 0) + 1)

Reviewer: Not sure how I feel about an example that could pretty much be copy-pasted.

Author: Oh. Yeah that's an issue. :( I should probably fix it.

Author: I've updated the example.

> While `getMode` is looking great, the `forEach` loop can simplified further. In order to do so, the function will be turned into an `arrow function expression`. This kind of expression is known for not using curly braces and shorter syntax. 
> 
> An example of such an expression is the following: 
> 
> ```js
> const numbersArr = [1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4];
> const counts = {};
> numbersArr.forEach(el => counts[el] = (counts[el] || 0) + 1)
> ```
> 
> In this example, the `numbersArr` has each element looped through like before. Only this time, it removes the if statement and replaces it with an `or` operator. This `or` operator says if `counts[el]` is falsy to return the other value. 
> 
> An value can be considered <dfn>falsy</dfn> if it is `null`, `undefined`, or otherwise can be evaluated as false. 
> 
> Either way the value of `counts[el]` gets incremented by one. Turn the existing `forEach` loop into an `arrow function expression`. 

Reviewer: ```suggestion
Your `forEach` loop is fully functional, but it's very verbose. You can actually simplify it further.

To do so, you will need to use an <dfn>arrow function expression</dfn>. This kind of function expression does not use curly braces, opting for a more compact single-line syntax.

Here is an example of an arrow function expression: 

```js
const numbersArr = [1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4];
const counts = {};
numbersArr.forEach(el => counts[el] = (counts[el] || 0) + 1)
```

In this example, `forEach` still iterates through every value in the array. The callback is now an arrow function expression.

The `(counts[el] || 0)` falls back to `0` when `counts[el]` is `undefined`, or another falsy value. This ensures that `counts[el]` will be incremented by 1, or initialised to 1 if the property does not have a truthy value.

Replace the callback in your `forEach` to an arrow function expression like the example.
```

---
id: 66bbb13f90639e7a6d664153
title: Step 34
challengeType: 0
dashedName: step-34
---

# --description--

While `getMode` is looking great, the `forEach` loop can simplified further. In order to do so, the function will be turned into an `arrow function expression`. This kind of expression is known for not using curly braces and shorter syntax. 

> An example of such a expression is the following: 

Reviewer: ```suggestion
An example of such an expression is the following: 
```

```js
const numbersArr = [1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4];
const counts = {};
numbersArr.forEach(el => counts[el] = (counts[el] || 0) + 1)
```

> In this example, the `numbersArr` has each element looped through like before. Only this time, it removes the 
> if statement and replaces it with an `or` operator. This `or` operator says if `counts[el]` is *falsy* to return the other value. 

Reviewer: ```suggestion
In this example, the `numbersArr` has each element looped through like before. Only this time, it removes the if statement and replaces it with an `or` operator. This `or` operator says if `counts[el]` is falsy to return the other value. 
```

> 

Reviewer: ```suggestion
```

> An value can be considered *falsy* if it is `null`, `undefined`, or otherwise can be evaluated as false. 

Reviewer: ```suggestion
An value can be considered <dfn>falsy</dfn> if it is `null`, `undefined`, or otherwise can be evaluated as false. 
```

Though I do feel like we've taught this before.

Author: We might have explained it in the shopping cart curriculum though it might not hurt to explain it multiple times.
